language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -0,0 +1,115 @@
+type WithMemory f = f TyName Name ExMemory
+
+newtype ExMemory = ExMemory Integer -- Counts size in machine words (64bit for the near future)
+  deriving (Eq, Ord, Show)
+  deriving newtype Num
+  deriving (Semigroup, Monoid) via (Sum Integer)
+newtype ExCPU = ExCPU Integer
+  deriving (Eq, Ord, Show)
+  deriving newtype Num
+  deriving (Semigroup, Monoid) via (Sum Integer)
+
+-- Based on https://github.com/ekmett/semigroups/blob/master/src/Data/Semigroup/Generic.hs
+class GExMemoryUsage f where
+  gmemoryUsage' :: f a -> ExMemory
+
+gmemoryUsage :: (Generic a, GExMemoryUsage (Rep a)) => a -> ExMemory
+gmemoryUsage x = gmemoryUsage' (from x)
+
+instance GExMemoryUsage U1 where
+  gmemoryUsage' _ = 1 -- No constructor
+
+instance GExMemoryUsage V1 where
+  gmemoryUsage' _ = 1 -- Empty datatype
+
+instance ExMemoryUsage a => GExMemoryUsage (K1 i a) where
+  gmemoryUsage' (K1 x) = memoryUsage x
+
+instance GExMemoryUsage f => GExMemoryUsage (M1 i c f) where
+  gmemoryUsage' (M1 x) = gmemoryUsage' x
+
+instance (GExMemoryUsage f, GExMemoryUsage g) => GExMemoryUsage (f :*: g) where
+  gmemoryUsage' (x1 :*: x2) = gmemoryUsage' x1 + gmemoryUsage' x2
+
+instance (GExMemoryUsage f, GExMemoryUsage g) => GExMemoryUsage (f :+: g) where
+  -- TODO I think this is supposed to count the max instead (?)
+  gmemoryUsage' (L1 x) = gmemoryUsage' x
+  gmemoryUsage' (R1 x) = gmemoryUsage' x
+
+newtype GenericExMemoryUsage a = GenericExMemoryUsage { getGenericExMemoryUsage :: a }
+instance (Generic a, GExMemoryUsage (Rep a)) => ExMemoryUsage (GenericExMemoryUsage a) where
+  memoryUsage (GenericExMemoryUsage x) = gmemoryUsage x
+
+class ExMemoryUsage a where
+    memoryUsage :: a -> ExMemory -- ^ How much memory does 'a' use?
+
+deriving via (GenericExMemoryUsage (Constant ann)) instance ExMemoryUsage ann => ExMemoryUsage (Constant ann)
+deriving via (GenericExMemoryUsage (Name ann)) instance ExMemoryUsage ann => ExMemoryUsage (Name ann)
+deriving via (GenericExMemoryUsage TypeBuiltin) instance ExMemoryUsage TypeBuiltin
+deriving via (GenericExMemoryUsage (Type TyName ann)) instance ExMemoryUsage ann => ExMemoryUsage (Type TyName ann)
+deriving via (GenericExMemoryUsage (Builtin ann)) instance ExMemoryUsage ann => ExMemoryUsage (Builtin ann)
+deriving via (GenericExMemoryUsage (Kind ann)) instance ExMemoryUsage ann => ExMemoryUsage (Kind ann)
+deriving via (GenericExMemoryUsage BuiltinName) instance ExMemoryUsage BuiltinName
+deriving via (GenericExMemoryUsage DynamicBuiltinName) instance ExMemoryUsage DynamicBuiltinName
+deriving via (GenericExMemoryUsage (Plain Term)) instance ExMemoryUsage (Plain Term)
+deriving via (GenericExMemoryUsage (WithMemory Term)) instance ExMemoryUsage (WithMemory Term)
+deriving newtype instance ExMemoryUsage ann => ExMemoryUsage (TyName ann)
+deriving newtype instance ExMemoryUsage ExMemory
+deriving newtype instance ExMemoryUsage Unique
+
+instance ExMemoryUsage () where
+  memoryUsage _ = 0 -- TODO or 1?
```

Comment:
We embed `() :: ()` as [`unitval`](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Unit.hs#L19), which is a PLC term. I'm not sure what our costing policy here is.

Reply:
This is also the base cost of a product type.